File: examples/example.android/app/src/main/java/io/oz/album/client/AlbumContext.java
package io.oz.album.client;

import android.content.SharedPreferences;
import android.content.res.Resources;

import java.security.GeneralSecurityException;

import io.odysz.common.LangExt;
import io.odysz.jclient.Clients;
import io.odysz.jclient.SessionClient;
import io.odysz.jclient.tier.ErrorCtx;
import io.odysz.semantic.jprotocol.AnsonMsg;
import io.odysz.semantic.jprotocol.JProtocol;
import io.oz.AlbumApp;
import io.oz.R;

public class AlbumContext {

    public boolean needSetup() {
        return LangExt.isblank(jserv, "/", ".", "http://", "https://")
                || LangExt.isblank(photoUser.device, "/", ".")
                || LangExt.isblank(photoUser.uid);
    }

    public enum ConnState { Online, Disconnected, LoginFailed }

    protected static final boolean verbose = true;

    public final String clientUri = "album.an";
    public final ErrorCtx errCtx = new ErrorCtx() {
        String msg;
        AnsonMsg.MsgCode code;
        @Override
        public void onError(AnsonMsg.MsgCode c, String msg, Object ... obj) {
            code = c;
            try {
                this.msg = String.format(msg, obj);
            }catch (Exception e) {
                this.msg = msg;
            }
        }
    };

    String jserv;
    String jdocbase;

    public AlbumClientier tier;

    public SessionClient client;

    public PhotoUser photoUser;

    ConnState state;
    public ConnState state() { return state; }

    public AlbumContext() {
        state = ConnState.Disconnected;
        photoUser = new PhotoUser(null);
    }

    public void init(Resources resources, SharedPreferences sharedPref) {
        jdocbase = resources.getString(R.string.jserv_docbase);

        photoUser.uid = sharedPref.getString(AlbumApp.keys.usrid, "");
        photoUser.pswd = sharedPref.getString(AlbumApp.keys.pswd, "");
        jserv = sharedPref.getString(AlbumApp.keys.jserv, "");
        photoUser.device = sharedPref.getString(AlbumApp.keys.device, "");

        Clients.init(jserv + "/" + jdocbase, verbose);
    }

    AlbumContext login(String uid, String pswd, TierCallback onOk, JProtocol.OnError onErr) throws GeneralSecurityException {
        // uid = "ody";
        // pswd = "123456";
        if (LangExt.isblank(photoUser.device, ".", "/"))
            throw new GeneralSecurityException("Device Id is null.");

        Clients.init(jserv + "/" + jdocbase, verbose);

        Clients.loginAsync(uid, pswd,
        (client) -> {
            tier = new AlbumClientier(clientUri, client, errCtx);
            state = ConnState.Online;
            if (onOk != null)
                onOk.ok(tier);
        },
        // Design Note: since error context don't have unified error message box,
        // error context pattern is not applicable.
        // errCtx.onError(c, r, (Object)v);
        (c, r, v) -> {
            state = ConnState.LoginFailed;
            if (onErr != null)
                onErr.err(c, r, v);
        }, photoUser.device);
        return this;
    }

    public void login(TierCallback onOk, JProtocol.OnError onErr)
            throws GeneralSecurityException {
        login(photoUser.uid(), photoUser.pswd(), onOk, onErr);
    }

    public AlbumContext jserv(String newVal) {
        jserv = newVal;
        Clients.init(jserv + "/" + jdocbase, verbose);
        return this;
    }

    public String jserv() { return jserv; }
}

File: examples/example.cs/album-wpf/album-sync/album/tier/AlbumContext.cs
using album_sync.album.tier;
using anclient.net.jserv.tier;
using io.odysz.anclient;
using io.odysz.anson.common;
using io.odysz.semantic.jprotocol;
using io.odysz.semantic.jsession;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Windows;
using static io.odysz.semantic.jprotocol.AnsonMsg;
using static io.oz.album.tier.AlbumContext;

namespace io.oz.album.tier
{
    public class PrefKeys
    {
        public const string device = "device";
        public const string jserv = "jserv";
        public const string port = "port";
        public const string logid = "logid";
    }

    class LoginHandler : OnLogin, JProtocol.OnError
    {
        public void err(MsgCode code, string msg, string[] args = null)
        {
            AlbumContext.GetInstance().State(ConnState.LoginFailed);
            AlbumContext.GetInstance().onError(code, msg, args);
        }

        /*
        public void ok(AnsonResp resp)
        {
            AlbumContext.GetInstance().loggedIn = true;
            AlbumContext.GetInstance().client = client;

                (SessionClient client) =>
                {
                    client.closeLink();

                    tier = new AlbumClientier(clientUri, client, errCtx);

                    client.openLink(clientUri, onHeartbeat, onLinkBroken, 19900); // 4 fails in 3 min
                            state = ConnState.Online;
                            if (onOk != null)
                                onOk.ok(tier);
                },
                // Design Note: since error context don't have unified error message box,
                // error context pattern of React is not applicable.
                // errCtx.onError(c, r, (Object)v);
                (MsgCode c, string r, object v) => {
                    state = ConnState.LoginFailed;
                    if (onErr != null)
                        onErr.err(c, r, v);
                }, photoUser.device);
        }
        */

        /// <summary>
        /// On login
        /// </summary>
        /// <param name="client"></param>
        public void ok(SessionClient client)
        {
            client.closeLink();
            AlbumClientier tier = new AlbumClientier(AlbumContext.clientUri, client, AlbumContext.GetInstance());
            client.openLink(AlbumContext.clientUri, heartbeatHandler, heartbeatHandler, 19900); // 4 fails in 3 min
            AlbumContext.GetInstance().State(ConnState.Online);

            // if (onOk != null) onOk.ok(tier);
        }

        HeartbeatHandler heartbeatHandler = new HeartbeatHandler();

    }

    internal class HeartbeatHandler : JProtocol.OnOk, JProtocol.OnError
    {
        public HeartbeatHandler()
        {
        }

        public void err(MsgCode code, string msg, string[] args = null)
        {
            if (AlbumContext.GetInstance().State() == ConnState.Disconnected)
                ; // how to notify?
            AlbumContext.GetInstance().State(ConnState.Online);
        }

        public void ok(AnsonResp resp)
        {
            AlbumContext.GetInstance().State(ConnState.Disconnected);
        }
    }

    public class AlbumContext : ErrorCtx
    {
        static AlbumContext instance;
        public static AlbumContext GetInstance()
        {
            if (instance == null)
                instance = new AlbumContext();
            return instance;
        }

        public SessionClient client { get; set; }
        public bool loggedIn { get; internal set; }

        protected bool verbose = true;

        public const string clientUri = "album.cs";
        public const string jdocbase  = "jserv-album";
        public const string albumHome = "dist/index.html";
        public const string synchPage = "dist/sync.html";

        public string jserv {
            get { return Preferences.jserv; }
            set {  Preferences.jserv = value; } 
        }
        
        public string pswd;
        public AlbumContext Pswd(string pswd)
        {
            this.pswd = pswd;
            return this;
        }


        public bool needSetup()
        {
            return LangExt.isblank(jserv, new string[] { "/", ".", "http://", "https://" })
                    || LangExt.isblank(photoUser.device, "/", ".")
                    || LangExt.isblank(photoUser.uid);
        }

        public enum ConnState { Online, Disconnected, LoginFailed }

        /*
        public const ErrorCtx errCtx = new ErrorCtx()
        {
            string msg;
            AnsonMsg.MsgCode code;

            public override void err(AnsonMsg.MsgCode c, string msg)
            {
                code = c;
                this.msg = msg;
            }
        };
        */

        public string homeName;

        public AlbumClientier tier;

        public SessionInf photoUser;

        ConnState state;
        public ConnState State() { return state; }
        public AlbumContext State(ConnState state) {
            this.state = state;
            return this;
        }

        public AlbumContext()
        {
            state = ConnState.Disconnected;

            // jserv = Preferences.jserv;
            Assert.IsNotNull(jserv);
        }

        AlbumContext login(string uid, string pswd, TierCallback onLogin, JProtocol.OnError onErr)
        {
            if (LangExt.isblank(photoUser.device, new string[] { "\\.", "/", "\\?", ":" }))
                throw new Exception("Device Id is null.");

            AnClient.Init(jserv + "/" + jdocbase, verbose);

            LoginHandler callback = new LoginHandler();
            AnClient.Login(uid, pswd, photoUser.device, callback, callback);

            /*
                (SessionClient client) =>
                {
                    client.closeLink();

                    tier = new AlbumClientier(clientUri, client, errCtx);

                    client.openLink(clientUri, onHeartbeat, onLinkBroken, 19900); // 4 times failed in 3 min
                            state = ConnState.Online;
                            if (onOk != null)
                                onOk.ok(tier);
                },
                // Design Note: since error context don't have unified error message box,
                // error context pattern of React is not applicable.
                // errCtx.onError(c, r, (Object)v);
                (MsgCode c, string r, object v) => {
                    state = ConnState.LoginFailed;
                    if (onErr != null)
                        onErr.err(c, r, v);
                }, photoUser.device);
            */
            return this;
        }

        public void login(TierCallback onOk, JProtocol.OnError onErr)
        {
            login(photoUser.uid, pswd, onOk, onErr);
        }

        /*
        JProtocol.OnOk onHeartbeat = ((resp) => {
                if (state == ConnState.Disconnected)
                    ; // how to notify?
            state = ConnState.Online;
        });

        JProtocol.OnError onLinkBroken = ((c, r, args) => {
                state = ConnState.Disconnected;
        });
        */

        public AlbumContext Jserv(string newVal)
        {
            jserv = newVal;
            AnClient.Init(jserv + "/" + jdocbase, verbose);
            return this;
        }

        #region ErrorCtx
        public override void onError(MsgCode code, string msg, string[] args = null)
        {
            MessageBox.Show(msg, "Error: " + code.Name(), MessageBoxButton.OK, MessageBoxImage.Error);
        }

        #endregion
    }
}
